File: packages/provide-uterm/src/provide/uterm/session_logger.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""JSONL session logger for recording BBS sessions."""

from __future__ import annotations

import asyncio
import base64
import contextlib
import json
import time
from pathlib import Path
from typing import TYPE_CHECKING, Any, Literal

from provide.telemetry import get_logger

if TYPE_CHECKING:
    from provide.uterm.recording import RecordingStore


logger = get_logger(__name__)


class SessionLogger:
    """Async session recorder using a pluggable RecordingStore.

    Each log entry is a JSON object on its own line with at minimum:
    ``{"ts": ..., "event": ..., "data": {...}}``.
    """

    def __init__(
        self,
        store: RecordingStore | str | Path,
        max_bytes: int = 0,
        *,
        control_channel_mode: Literal["exclude", "wire"] = "exclude",
        flush_interval_s: float = 5.0,
        batch_size: int = 100,
    ) -> None:

        self._log_path: Path | None
        if isinstance(store, (str, Path)):
            p = Path(store)
            # Legacy compatibility: tests pass a full file path like tmp/s.jsonl.
            # We must ensure that start(session_id) writes to THIS EXACT path,
            # regardless of session_id.
            from provide.uterm.recording import RecordingStore

            class LegacyFileStore(RecordingStore):
                def __init__(self, path: Path):
                    self._path = path

                async def start_session(self, session_id: str, metadata: dict[str, Any]) -> None:
                    self._path.parent.mkdir(parents=True, exist_ok=True)
                    event = {"ts": time.time(), "event": "log_start", "data": metadata, "session_id": session_id}
                    # Always append to support 'test_file_opens_in_append_mode'
                    with self._path.open("a", encoding="utf-8") as f:
                        f.write(json.dumps(event) + "\n")

                async def append_events(self, session_id: str, events: list[dict[str, Any]]) -> None:
                    with self._path.open("a", encoding="utf-8") as f:
                        for e in events:
                            f.write(json.dumps(e) + "\n")

                async def end_session(self, session_id: str) -> None:
                    event = {"ts": time.time(), "event": "log_stop", "data": {}, "session_id": session_id}
                    with self._path.open("a", encoding="utf-8") as f:
                        f.write(json.dumps(event) + "\n")

                async def get_path(self, session_id: str) -> Path | None:
                    return self._path

                async def recording_meta(self, session_id: str) -> dict[str, Any]:
                    return {
                        "session_id": session_id,
                        "exists": self._path.exists(),
                        "size_bytes": self._path.stat().st_size if self._path.exists() else 0,
                    }

                async def get_entries(
                    self,
                    session_id: str,
                    limit: int = 200,
                    offset: int | None = None,
                    event: str | None = None,
                ) -> list[dict[str, Any]]:
                    if not self._path.exists():
                        return []
                    normalized_limit = max(1, min(limit, 500))
                    entries: list[dict[str, Any]] = []
                    skipped = 0
                    with self._path.open(encoding="utf-8") as f:
                        for line in f:
                            try:
                                item = json.loads(line)
                            except json.JSONDecodeError:
                                continue
                            if event and item.get("event") != event:
                                continue
                            if offset is not None and skipped < offset:
                                skipped += 1
                                continue
                            entries.append(item)
                    if offset is None:
                        return entries[-normalized_limit:]
                    return entries[:normalized_limit]

            self._store: RecordingStore = LegacyFileStore(p)
            self._log_path = p
        else:
            self._store = store
            self._log_path = None

        self._lock = asyncio.Lock()
        self._session_id: str | None = None
        self._context: dict[str, str] = {}
        self._max_bytes = max_bytes  # 0 = unlimited
        self._control_channel_mode = control_channel_mode
        self._bytes_written = 0
        self._quota_warned = False
        self._batch_size = batch_size
        self._flush_interval = flush_interval_s
        self._buffer: list[dict[str, Any]] = []
        self._flush_task: asyncio.Task[None] | None = None

    @property
    def _file(self) -> Any:
        # Legacy compatibility for tests checking if file is closed
        return None

    async def _write_event_unlocked(self, event: str, data: dict[str, Any]) -> None:
        # Legacy compatibility for tests mocking this
        await self._write_event(event, data)

    async def start(self, session_id: str) -> None:
        """Begin a recording session."""
        self._session_id = session_id
        metadata: dict[str, Any] = {"started_at": time.time()}
        if self._log_path:
            metadata["path"] = str(self._log_path)
        await self._store.start_session(session_id, metadata)

        # Initialize bytes_written from store metadata
        meta = await self._store.recording_meta(session_id)
        try:
            self._bytes_written = int(meta.get("size_bytes", 0))
        except (TypeError, ValueError):
            self._bytes_written = 0

        self._flush_task = asyncio.create_task(self._periodic_flush())

    async def stop(self) -> None:
        """Finalize the recording session."""
        if self._flush_task:
            self._flush_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._flush_task

        await self._flush_buffer()
        if self._session_id:
            await self._store.end_session(self._session_id)

    async def log_send(self, keys: str) -> None:
        """Log sent keystrokes."""
        payload = keys.encode("cp437", errors="replace")
        await self._write_event("send", {"keys": keys, "bytes_b64": base64.b64encode(payload).decode("ascii")})

    async def log_send_masked(self, byte_count: int) -> None:
        """Log a credential send without capturing the actual value."""
        await self._write_event(
            "send",
            {
                "keys": "***",
                "bytes_b64": base64.b64encode(b"***").decode("ascii"),
                "masked": True,
                "byte_count": byte_count,
            },
        )

    async def log_screen(self, snapshot: dict[str, Any], raw: bytes) -> None:
        """Log a screen snapshot with raw bytes."""
        data = {
            **snapshot,
            "raw": raw.decode("cp437", errors="replace"),
            "raw_bytes_b64": base64.b64encode(raw).decode("ascii"),
        }
        await self._write_event("read", data)

    async def log_event(self, event: str, data: dict[str, Any]) -> None:
        """Log an arbitrary named event."""
        await self._write_event(event, data)

    async def log_wire(self, direction: Literal["send", "recv"], text: str) -> None:
        """Log a raw wire chunk when wire-mode recording is enabled."""
        if self._control_channel_mode != "wire":
            return
        payload = text.encode("utf-8")
        await self._write_event(
            f"wire_{direction}",
            {
                "text": text,
                "bytes_b64": base64.b64encode(payload).decode("ascii"),
            },
        )

    async def log_control(self, direction: Literal["send", "recv"], control: dict[str, Any]) -> None:
        """Log a decoded control frame when wire-mode recording is enabled."""
        if self._control_channel_mode != "wire":
            return
        await self._write_event(f"control_{direction}", {"control": control})

    def set_context(self, context: dict[str, str]) -> None:
        """Set metadata context for subsequent log entries."""
        self._context = {str(k): str(v) for k, v in context.items()}

    def clear_context(self) -> None:
        """Clear metadata context."""
        self._context = {}

    async def flush(self) -> None:
        """Manually flush buffered log entries."""
        await self._flush_buffer()

    async def _write_event(self, event: str, data: dict[str, Any]) -> None:
        async with self._lock:
            if self._max_bytes > 0 and self._bytes_written >= self._max_bytes:
                if not self._quota_warned:
                    self._quota_warned = True
                    logger.warning("session_logger_quota_reached — further writes suppressed")
                return

            record: dict[str, Any] = {"ts": time.time(), "event": event, "data": data}
            if self._session_id:
                record["session_id"] = self._session_id
            if self._context:
                record["ctx"] = dict(self._context)

            self._buffer.append(record)
            # Better estimate: actual JSON size + newline
            self._bytes_written += len(json.dumps(record)) + 1

            if len(self._buffer) >= self._batch_size:
                await self._flush_buffer_unlocked()

    async def _flush_buffer_unlocked(self) -> None:
        if not self._buffer or not self._session_id:
            return
        batch = list(self._buffer)
        self._buffer.clear()
        await self._store.append_events(self._session_id, batch)

    async def _flush_buffer(self) -> None:
        async with self._lock:
            await self._flush_buffer_unlocked()

    async def _periodic_flush(self) -> None:
        while True:
            await asyncio.sleep(self._flush_interval)
            await self._flush_buffer()

File: packages/provide-uterm-csharp/src/Provide.Uterm/SessionLogger/SessionLogger.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Text;
using Provide.Uterm.Recording;
using Provide.Uterm.Redaction;

namespace Provide.Uterm.SessionLogger;

public enum ControlChannelMode
{
    Exclude,
    Wire,
}

public sealed class SessionLoggerOptions
{
    public int MaxBytes { get; set; }
    public ControlChannelMode ControlChannelMode { get; set; } = ControlChannelMode.Exclude;
    public Redactor? Redactor { get; set; }
    public TimeSpan FlushInterval { get; set; } = TimeSpan.FromSeconds(5);
    public int BatchSize { get; set; } = 100;
}

/// <summary>
/// Async session recorder over a pluggable recording store.
/// Port of provide.uterm.session_logger / packages/provide-uterm-go/sessionlogger.
/// </summary>
public sealed class SessionLogger : IAsyncDisposable
{
    private readonly IRecordingStore _store;
    private readonly int _maxBytes;
    private readonly ControlChannelMode _controlChannelMode;
    private readonly Redactor? _redactor;
    private readonly TimeSpan _flushInterval;
    private readonly int _batchSize;

    private readonly object _lock = new();
    private string _sessionId = "";
    private int _bytesWritten;
    private readonly List<Event> _buffer = new();
    private CancellationTokenSource? _flushCts;
    private Task? _flushTask;

    public SessionLogger(IRecordingStore store, SessionLoggerOptions? options = null)
    {
        options ??= new SessionLoggerOptions();
        _store = store;
        _maxBytes = options.MaxBytes;
        _controlChannelMode = options.ControlChannelMode;
        _redactor = options.Redactor;
        _flushInterval = options.FlushInterval <= TimeSpan.Zero
            ? TimeSpan.FromSeconds(5)
            : options.FlushInterval;
        _batchSize = options.BatchSize <= 0 ? 100 : options.BatchSize;
    }

    public async Task StartAsync(string sessionId)
    {
        lock (_lock)
        {
            _sessionId = sessionId;
        }

        var metadata = new Dictionary<string, object?>
        {
            ["started_at"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0,
        };
        await _store.StartSessionAsync(sessionId, metadata);
        var meta = await _store.RecordingMetaAsync(sessionId);
        lock (_lock)
        {
            _bytesWritten = (int)meta.SizeBytes;
            _flushCts = new CancellationTokenSource();
            _flushTask = PeriodicFlushAsync(_flushCts.Token);
        }
    }

    public async Task StopAsync()
    {
        CancellationTokenSource? cts;
        Task? task;
        lock (_lock)
        {
            cts = _flushCts;
            task = _flushTask;
            _flushCts = null;
            _flushTask = null;
        }

        if (cts is not null)
        {
            await cts.CancelAsync();
            if (task is not null)
            {
                try
                {
                    await task;
                }
                catch (OperationCanceledException)
                {
                }
            }

            cts.Dispose();
        }

        await FlushAsync();
        string sessionId;
        lock (_lock)
        {
            sessionId = _sessionId;
        }

        if (sessionId.Length > 0)
        {
            await _store.EndSessionAsync(sessionId);
        }
    }

    public async Task LogAsync(string eventName, IReadOnlyDictionary<string, object?> data)
    {
        if (_controlChannelMode == ControlChannelMode.Exclude &&
            (eventName is "wire" or "control"))
        {
            return;
        }

        var payload = new Dictionary<string, object?>(data);
        if (_redactor is not null)
        {
            foreach (var key in payload.Keys.ToList())
            {
                if (payload[key] is string s)
                {
                    payload[key] = Redaction.Redaction.RedactText(s, _redactor);
                }
            }
        }

        string sessionId;
        lock (_lock)
        {
            sessionId = _sessionId;
            if (_maxBytes > 0 && _bytesWritten >= _maxBytes)
            {
                return;
            }

            var evt = new Event
            {
                ["ts"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0,
                ["event"] = eventName,
                ["data"] = payload,
                ["session_id"] = sessionId,
            };
            _buffer.Add(evt);
            if (_buffer.Count >= _batchSize)
            {
                // fall through to flush
            }
            else
            {
                return;
            }
        }

        await FlushAsync();
    }

    public async Task FlushAsync()
    {
        List<Event> batch;
        string sessionId;
        lock (_lock)
        {
            if (_buffer.Count == 0)
            {
                return;
            }

            batch = _buffer.ToList();
            _buffer.Clear();
            sessionId = _sessionId;
        }

        if (sessionId.Length == 0)
        {
            return;
        }

        await _store.AppendEventsAsync(sessionId, batch);
        lock (_lock)
        {
            foreach (var e in batch)
            {
                _bytesWritten += Encoding.UTF8.GetByteCount(System.Text.Json.JsonSerializer.Serialize(e)) + 1;
            }
        }
    }

    private async Task PeriodicFlushAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                await Task.Delay(_flushInterval, ct);
            }
            catch (OperationCanceledException)
            {
                break;
            }

            await FlushAsync();
        }
    }

    public async ValueTask DisposeAsync() => await StopAsync();
}
